File: 2019 카카오 개발자 겨울 인턴십/호텔 방 배정.java
import java.util.*;

class Solution {
    Map<Long, Long> room = new HashMap<>();
    
    long find(long n) {
        if(room.containsKey(n) == false) return n;
        room.put(n, find(room.get(n)));
        return room.get(n);
    }
    
    public long[] solution(long k, long[] room_number) {
        int n = room_number.length;
        long[] answer = new long[n];
        for(int i = 0; i < n; ++i) {
            long num = room_number[i];
            if(room.containsKey(num) == false) {
                answer[i] = num;
                room.put(num, find(num+1));
            }
            else {
                long nextNum = find(num);
                answer[i] = nextNum;
                room.put(nextNum, find(nextNum+1));
            }
        }
        return answer;
    }
}

File: 2019 카카오 개발자 겨울 인턴십/호텔 방 배정.cpp
#include <iostream>
#include <string>
#include <vector>
#include <map>

using namespace std;

map<long long, long long> room;

long long find(long long n) {
    if(room[n] == 0) return n;
    return room[n] = find(room[n]);
}

vector<long long> solution(long long k, vector<long long> room_number) {
    vector<long long> answer;
    for(int i = 0; i < room_number.size(); ++i) {
        long long num = room_number[i];
        if(room[num] == 0) {
            answer.push_back (num);
            room[num] = find(num+1);
        }
        else {
            long long nextNum = find(num);
            answer.push_back(nextNum);
            room[nextNum] = find(nextNum+1);
        }
    }
    for(auto it = room.begin(); it != room.end(); ++it)
        cout << it->first << " " << it->second << endl;
    return answer;
}
